packages/wrangler/src/deploy/config-diffs.ts

```diff
@@ -0,0 +1,296 @@
+import { green, red } from "@cloudflare/cli/colors";
+import { Diff } from "../cloudchamber/helpers/diff";
+import type { Config, RawConfig } from "../config";
+
+/**
+ * Object representing the difference of two configuration objects.
+ */
+type ConfigDiff = {
+	/** The actual (raw) computed diff of the two objects */
+	diff: Diff;
+	/** Boolean indicating whether the difference is not present or only comprized of the addition of configuration options */
+	onlyAdditionsIfAny: boolean;
+};
+
+/**
+ * Computes the difference between a remote representation of a Worker's config and a local configuration.
+ *
+ * @param remoteConfig The remote representation of a Worker's config
+ * @param localConfig The local config
+ * @param localResolvedConfig The resolved local config (used to resolve some values on the raw config)
+ * @returns Object containing the diffing information
+ */
+export function getRemoteConfigDiff(
+	remoteConfig: RawConfig,
+	localConfig: RawConfig,
+	localResolvedConfig: Config
+): ConfigDiff {
+	const normalizedLocal = normalizeLocalConfigAsRemote(
+		localConfig,
+		localResolvedConfig
+	);
+
+	// We reorder the remote config so that its ordering follows that
+	// of the local one (this ensures that the diff users see lists
+	// the configuration options in the same order as their config file)
+	const reorderedRemoteConfig: RawConfig = orderObjectFields(
+		remoteConfig as Record<string, unknown>,
+		localConfig as Record<string, unknown>
+	);
+
+	const diff = new Diff(
+		JSON.stringify(reorderedRemoteConfig, null, 2),
+		JSON.stringify(normalizedLocal, null, 2)
+	);
+
+	return {
+		diff,
+		onlyAdditionsIfAny: configDiffOnlyHasAdditionsIfAny(diff),
+	};
+}
+
+/**
+ * Given a diff object, representing the diff of config files, it computes whether such diff
+ * object only represents at most additions (so no removal nor modifications).
+ *
+ * For example the diff:
+ *  ```
+ *   - "name": "my-worker"
+ *   + "name": "my-worker",
+ *   + "vars": {
+ *   +   MY_VAR: "my variable",
+ *   + },
+ *  ```
+ * only contains additions.
+ *
+ * While
+ *  ```
+ *    "name": "my-worker",
+ *    "vars": {
+ *   -   MY_VAR: "my variable",
+ *   +   MY_VAR: "my modified variable",
+ *    },
+ *	 - compatibility_date: "2025-07-08",
+ *  ```
+ * contains also a modification and a removal
+ *
+ * @param diff The diff object to analyze
+ * @returns true if there are only additions or no diffs at all, false otherwise
+ */
+function configDiffOnlyHasAdditionsIfAny(diff: Diff): boolean {
+	const diffLines = `${diff}`.split("\n");
```

Comment:
Is this implicitly calling `diff.toString()`?

Reply:
updated